### PR TITLE
OpenTelemetry Tracing for ProductCatalogService

### LINKED ISSUE #248: OpenTelemetry Tracing for ProductCatalogService
Subtask of #132 
Use OpenTelemetry for tracing in the productcatalog service.

### PR DESCRIPTION
**Pull request by gzhao408**
Resolves #248 . Subtask of #132 .

To test, deploy the sandbox and you should see this in Cloud Trace (specifically, look for Span.{server, client}-hipstershop.ProductCatalogService/{ListProducts, GetProduct}):

image

image

### CODE REVIEW

File: src/frontend/main.go [hunk 1]
```diff
@@ -120,12 +120,12 @@ func main() {
 	go initTelemetry(log)
 
 	// TODO: register views when OpenTelemetry Views API is available
-	meter := controller.Provider().Meter("hipstershop/frontend")
+	meter := controller.Provider().Meter("hipstershop")
```
Comment (simonz130) [inline src/frontend/main.go:123]: Is this change related to product catalog service?

File: src/productcatalogservice/server.go [hunk 2]
```diff
@@ -78,7 +85,8 @@ func init() {
 }
 
 func main() {
-	initStackDriverTracing()
+	initOpenCensus()
```
Comment (simonz130) [inline src/productcatalogservice/server.go:88]: Maybe rename OpenCensus method to be more concrete: initOpenCensusMetrics
This way it's more clear what is part of OpenTelemetry vs OpenCensus instrumentation as we work towards getting rid of OpenCensus completely.

File: src/productcatalogservice/server.go [hunk 3]
```diff
@@ -119,14 +127,20 @@ func run(port int) string {
...
 	svc := &productCatalog{}
 	pb.RegisterProductCatalogServiceServer(srv, svc)
 	healthpb.RegisterHealthServer(srv, svc)
 	go srv.Serve(l)
 	return l.Addr().String()
 }
 
+// Initialize OpenCensus Stats (Metrics)
+// TODO: remove this after full conversion to OpenTelemetry
 func initStats(exporter *stackdriver.Exporter) {
```
Comment (simonz130) [inline src/productcatalogservice/server.go:144]: Make it clear it's OpenCensus in the method name

File: src/frontend/main.go (hunk 1 above)
Comment (gzhao408) [reply src/frontend/main.go]: removed this change

File: src/productcatalogservice/server.go (hunk 2 above)
Comment (gzhao408) [reply src/productcatalogservice/server.go]: I just moved the helper function in there and called it initOpenCensusStats, and commented that this would be replaced with OpenTelemetry Metrics.

**Comment by gzhao408** [comment]
@simonz130 did a [refactor](https://github.com/GoogleCloudPlatform/stackdriver-sandbox/compare/c5bf1ff936c873780074a7705681f7971e3b98e9..95db4d2c0be67909d36ee8ac71a894184f6d5501)

**Comment by Daniel-Sanche** [comment]
For some reason I'm not seeing those traces, only BatchWriteSpans and health checks. I'm not too familiar with Cloud Trace though, so it could just be a me problem

**Comment by gzhao408** [comment]
> For some reason I'm not seeing those traces

:O no span.server-frontend whatsoever? Oh, maybe you need to open hipstershop and click around a few times, or have the loadgenerator hit it with some requests, otherwise there might not be anything to trace?

File: src/productcatalogservice/server.go [hunk 4]
```diff
@@ -164,6 +169,27 @@ func initStackDriverTracing() {
 	log.Warn("could not initialize stackdriver exporter after retrying, giving up")
 }
 
+// Initialize OTel trace provider that exports to Cloud Trace
+func initTraceProvider() {
+	// Initialize exporter OTel Trace -> Google Cloud Trace
+	projectID := os.Getenv("GOOGLE_CLOUD_PROJECT")
+	exporter, err := texporter.NewExporter(texporter.WithProjectID(projectID))
```
Comment (simonz130) [inline src/productcatalogservice/server.go:176]: Make sure the projectId is not null or empty as well as check the presense of the GOOGLE_APPLICATION_CREDENTIALS env variable to make sure that auth will work.

File: src/productcatalogservice/server.go [hunk 5]
```diff
@@ -164,6 +169,27 @@ func initStackDriverTracing() {
...
+	}
+
+	// Create trace provider with the exporter.
+	tp, err := sdktrace.NewProvider(sdktrace.WithConfig(
+		sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}),
+		sdktrace.WithSyncer(exporter),
+		// TODO: replace with predefined constant for GKE or autodetection when available
+		sdktrace.WithResource(resource.New(standard.ServiceNameKey.String("GKE"))))
+	if err != nil {
+		log.Fatal("failed to initialize trace provider: %v", err)
```
Comment (simonz130) [inline src/productcatalogservice/server.go:188]: I don't think we should fail the app if tracing doesn't work, but we should log it as critical error.

**Comment by Daniel-Sanche** [comment]
> :O no span.server-frontend whatsoever? Oh, maybe you need to open hipstershop and click around a few times, or have the loadgenerator hit it with some requests, otherwise there might not be anything to trace?

Nevermind, I tried again today and it's working!

File: src/productcatalogservice/server.go (hunk 4 above)
Comment (gzhao408) [reply src/productcatalogservice/server.go]: checking that projectID isn't empty, I don't think it's necessary to check credentials, would be the responsibility of the [exporter](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/4bf57be619cd46e7dfce46107fc7baa368ebc9ef/exporter/trace/README.md#authentication)

File: src/productcatalogservice/server.go (hunk 5 above)
Comment (gzhao408) [reply src/productcatalogservice/server.go]: yep, got it :)

**Comment by gzhao408** [comment]
@simonz130 Please take another look, I'm requiring PROJECT_ID since it's a required parameter, but the auth/credentials are the [responsibility of the exporter](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/4bf57be619cd46e7dfce46107fc7baa368ebc9ef/exporter/trace/README.md#authentication). I'm having it retry up to 3 times and just warn if it fails. Thanks!
